eis_toolkit/prediction/cnn_classification_and_probability.py

```diff
@@ -0,0 +1,353 @@
+       neuron_list: List of unit or neuron used to build the network final part of the network. the length of the list
+                    is the number of fully connected layers.
+       dropout_rate: Float number that help avoiding over-fitting. It just randomly drops samples.
+       output_units: number of output classes.
+       last_activation: usually you should use softmax or sigmoid.
+
+    Return:
+        model: the model that has been created.
+
+    Raises:
+        InvalidParameterValueException: Raised when the input is not valid.
+        InvalidParameterValueException: Raised when parameters like dropout is invalid.
+    """
+
+    # check that the input is not null
+    if input_shape_for_cnn is None:
+        raise InvalidParameterValueException
+
+    if len(neuron_list) <= 0:
+        raise InvalidArgumentTypeException
+
+    if dropout_rate is not None and dropout_rate <= 0:
+        raise InvalidParameterValueException
+
+    # generate the input
+    input_layer = tf.keras.Input(shape=input_shape_for_cnn)
+
+    if data_augmentation is not False:
+        input_layer = tf.keras.layers.RandomRotation((-0.2, 0.5))(input_layer)
+
+    # we do dynamically the building of the conv2d layers
+    for conv_layer_number, neuron in enumerate(conv_list):
+        if conv_layer_number == 0:
+            x = tf.keras.layers.Conv2D(
+                filters=neuron,
+                activation="relu",
+                padding="same",
+                kernel_regularizer=regularization,
+                kernel_size=convolution_kernel_size,
+            )(input_layer)
+        else:
+            x = tf.keras.layers.Conv2D(
+                filters=neuron,
+                activation="relu",
+                padding="same",
+                kernel_regularizer=regularization,
+                kernel_size=convolution_kernel_size,
+            )(x)
+
+        if dropout_rate is not None:
+            x = tf.keras.layers.Dropout(dropout_rate)(x)
+
+        x = tf.keras.layers.BatchNormalization()(x)
+        x = tf.keras.layers.MaxPool2D(pool_size=pool_size)(x)
+
+    for layer_number, neuron in enumerate(neuron_list):
+        if layer_number == 0:
+            body = tf.keras.layers.Dense(neuron, activation="relu", kernel_regularizer=regularization)(input_layer)
+        else:
+            body = tf.keras.layers.Dense(neuron, kernel_regularizer=regularization, activation="relu")(body)
+
+        if dropout_rate is not None:
+            body = tf.keras.layers.Dropout(dropout_rate)(body)
+
+    body = tf.keras.layers.Flatten()(body)
+
+    classifier = tf.keras.layers.Dense(output_units, activation=last_activation, kernel_regularizer=regularization)(
+        body
+    )
+
+    model = tf.keras.Model(inputs=input_layer, outputs=classifier)
+    model.compile(optimizer=optimizer, loss=loss, metrics=["accuracy"])
+    return model
+
+
+def train_and_predict_for_classification(
```

Comment:
This shouldnt be called "train and predict". Prediction should be a separate things, done on *unseen* data. Instead, you should just train the model here and return the validation metrics. The user can then test and predict using separate functions.

Reply:
yes yes of course! I do it!  My function do both train with the training data and predict with validation data I predict with unseen data I rename it run inference_for_classification!